tpm/tpm.go

```diff
@@ -0,0 +1,183 @@
+package tpm
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"sync"
+
+	"github.com/google/go-attestation/attest"
+
+	"go.step.sm/crypto/tpm/internal/open"
+	"go.step.sm/crypto/tpm/simulator"
+	"go.step.sm/crypto/tpm/storage"
+)
+
+// TPM models a Trusted Platform Module. It provides an abstraction
+// over the google/go-tpm and google/go-attestation packages, allowing
+// functionalities of these packages to be performed in a uniform manner.
+// Besides that, it provides a transparent method for persisting TPM
+// objects, so that referencing and using these is simplified.
+type TPM struct {
+	deviceName   string
+	attestConfig *attest.OpenConfig
+	attestTPM    *attest.TPM
+	rwc          io.ReadWriteCloser
+	lock         sync.RWMutex
+	store        storage.TPMStore
+	simulator    *simulator.Simulator
+	downloader   *downloader
+	info         *Info
+	eks          []*EK
+}
+
+// NewTPMOption is used to provide options when instantiating a new
+// instance of TPM.
+type NewTPMOption func(t *TPM) error
+
+// WithDeviceName is used to provide the `name` or path to the TPM
+// device.
+func WithDeviceName(name string) NewTPMOption {
+	return func(t *TPM) error {
+		t.deviceName = name
+		return nil
+	}
+}
+
+// WithStore is used to set the TPMStore implementation to use for
+// persisting TPM objects, including AKs and Keys.
+func WithStore(store storage.TPMStore) NewTPMOption {
+	return func(t *TPM) error {
+		if store == nil {
+			store = storage.BlackHole() // prevent nil storage; no persistence
+		}
+
+		t.store = store
+		return nil
+	}
+}
+
+// WithDisableDownload disables EK certificates from being downloaded
+// from online hosts.
+func WithDisableDownload() NewTPMOption {
+	return func(t *TPM) error {
+		t.downloader.enabled = false
+		return nil
+	}
+}
+
+// New creates a new TPM instance. It takes `opts` to configure
+// the instance.
+func New(opts ...NewTPMOption) (*TPM, error) {
+	tpm := &TPM{
+		attestConfig: &attest.OpenConfig{TPMVersion: attest.TPMVersion20},                      // default configuration for TPM attestation use cases
+		store:        storage.BlackHole(),                                                      // default storage doesn't persist anything // TODO(hs): make this in-memory storage instead?
+		downloader:   &downloader{enabled: true, maxDownloads: 10, client: http.DefaultClient}, // EK certificate download (if required) is enabled by default
+	}
+
+	for _, o := range opts {
+		if err := o(tpm); err != nil {
+			return nil, err
+		}
+	}
+
+	return tpm, nil
+}
+
+// Open readies the TPM for usage and marks it as being
+// in use. This makes using the instance safe for
+// concurrent use.
+func (t *TPM) open(ctx context.Context) error {
+	// prevent opening the TPM multiple times if Open is called
+	// within the package multiple times.
+	if isInternalCall(ctx) {
+		return nil
+	}
+
+	// lock the TPM instance; it's in use now
+	t.lock.Lock()
```

Comment:
perhaps we should unlock on errors.

Reply:
I added a `defer` with a check to do so.